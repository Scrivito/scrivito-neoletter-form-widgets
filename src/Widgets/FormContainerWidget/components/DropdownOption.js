import * as React from "react";

export const DropdownOption = ({ value, id }) => {
  return (
    <option value={value} id={id}>
      {value}
    </option>
  );
};
