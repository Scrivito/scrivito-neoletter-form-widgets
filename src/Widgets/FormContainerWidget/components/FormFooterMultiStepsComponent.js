import * as React from "react";
import * as Scrivito from "scrivito";

export const FormFooterMultiSteps = Scrivito.connect(
  ({
    widget,
    onPageChange,
    onSubmit,
    currentStep,
    isLastPage,
    stepsLength,
  }) => {
    return (
      <div className="form-buttons">
        <button
          className="btn btn-primary backward-button"
          onClick={() => onPageChange(false)}
          hidden={currentStep == 1 && !Scrivito.isInPlaceEditingActive()}
        >
          {widget.get("backwardButtonText")}
        </button>
        <div className="step-counter">{currentStep + " / " + stepsLength}</div>
        <button
          className="btn btn-primary forward-button"
          onClick={isLastPage ? onSubmit : () => onPageChange(true)}
        >
          {isLastPage
            ? widget.get("submitButtonText")
            : widget.get("forwardButtonText")}
        </button>
      </div>
    );
  }
);
