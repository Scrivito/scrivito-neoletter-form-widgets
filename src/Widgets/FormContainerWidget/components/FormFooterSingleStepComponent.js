import * as React from "react";
import * as Scrivito from "scrivito";

export const FormFooterSingleStep = Scrivito.connect(({ widget, onSubmit }) => {
  return (
    <div
      className={`${
        widget.get("singleSubmitButtonAlignment") === "block"
          ? ""
          : widget.get("singleSubmitButtonAlignment")
      }`}
    >
      <button
        className={`btn btn-primary${
          widget.get("singleSubmitButtonAlignment") === "block"
            ? " btn-block"
            : ""
        }`}
        onClick={onSubmit}
      >
        {widget.get("submitButtonText")}
      </button>
    </div>
  );
});
