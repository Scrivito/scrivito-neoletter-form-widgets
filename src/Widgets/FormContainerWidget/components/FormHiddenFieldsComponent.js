import * as React from "react";
import * as Scrivito from "scrivito";
import { getHistory } from "../../../config/history";

export const FormHiddenFields = ({ widget }) => {
  const [browserLocation, setBrowserLocation] = React.useState(null);
  React.useEffect(() => {
    const history = getHistory();
    console.log("And the history is....");
    console.log(history);
    if (!history) return;
    setBrowserLocation(locationToUrl(history.location));

    return history.listen(({ location }) =>
      setBrowserLocation(locationToUrl(location))
    );
  }, []);
  return (
    <>
      <input type="hidden" name="form_id" value={widget.get("formId")} />
      <input
        type="hidden"
        name="url"
        value={browserLocation || Scrivito.urlFor(widget.obj())}
      />
      <Scrivito.InPlaceEditingOff>
        <Scrivito.ContentTag content={widget} attribute="hiddenFields" />
      </Scrivito.InPlaceEditingOff>

      <HoneypotField />
    </>
  );
};

const HoneypotField = () => (
  <div aria-hidden="true" className="winnie-the-pooh">
    <input autoComplete="off" name="fax" tabIndex="-1" />
  </div>
);

function locationToUrl(location) {
  return `${window.location.origin}${location.pathname}${location.search}`;
}
