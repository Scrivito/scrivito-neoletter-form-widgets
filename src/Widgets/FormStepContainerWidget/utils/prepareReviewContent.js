import { getFieldName } from "./getFieldName";
import { map, uniq } from "lodash-es";

export function prepareReviewContent(widget) {
  const form = document.getElementById(widget.get("formId"));
  const data = new FormData(form);
  const joinedFormData = new FormData();
  // show all field-names with equal name as a comma separated string
  for (const [name, value] of data) {
    if (joinedFormData.has(name)) {
      continue;
    } else {
      joinedFormData.set(name, data.getAll(name).join(", "));
    }
  }

  const steps = widget.get("steps");
  const widgets = steps.flatMap((s) => s.widgets());
  const includeEmptyAnswers = widget.get("includeEmptyAnswers");
  const inputs = Array.from(form.querySelectorAll("input, select, textarea"));
  const inputNames = uniq(map(inputs, (i) => i.name));
  const reviewData = [];

  for (const key of inputNames) {
    const answer = joinedFormData.get(key);
    if (answer == undefined || answer == "") {
      // do not show empty answers
      if (!includeEmptyAnswers) {
        continue;
      }
    }
    // check if is hidden
    const input = inputs.find((i) => i.name == key);
    if (
      key == "fax" ||
      (input.type == "hidden" && !input.classList.contains("show-in-review"))
    ) {
      // do not show hidden inputs
      continue;
    }
    // get title && step number
    const widget = widgets.find((w) => getFieldName(w) == key);
    if (widget) {
      const step = steps.find((s) =>
        s.widgets().find((w) => getFieldName(w) == key)
      );
      const stepNumber = step.get("stepNumber");
      const title = widget.get("label") || widget.get("title") || "";

      if (!reviewData[stepNumber]) {
        reviewData[stepNumber] = [];
      }
      reviewData[stepNumber].push({
        title: title,
        value: getValue(answer, widget),
      });
    }
  }
  return reviewData;
}

function getValue(answer, widget) {
  const emptyValue = "-";
  if (!answer || answer.length <= 0) {
    return emptyValue;
  }
  const className = widget.objClass();
  if (className != "FormDateWidget") {
    return answer;
  }
  const type = widget.get("dateType");
  if (type == "date") {
    return new Date(answer).toLocaleDateString();
  } else if (type == "datetime-local") {
    return new Date(answer).toLocaleString();
  }
  return answer;
}
