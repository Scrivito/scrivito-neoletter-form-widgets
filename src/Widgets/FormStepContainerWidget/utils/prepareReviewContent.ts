import { isEmpty, map, uniq } from "lodash-es";
import { Widget } from "scrivito";
import { getFieldName } from "./getFieldName";
import { InputValidationElement, ReviewContent } from "../../../../types/types";

export function prepareReviewContent(widget: Widget): ReviewContent {
  const form = document.getElementById(
    widget.get("formId") as string
  ) as HTMLFormElement;
  const data = new FormData(form);
  const joinedFormData = new FormData();
  // show all field-names with equal name as a comma separated string
  for (const [name] of data) {
    if (joinedFormData.has(name)) {
      continue;
    } else {
      joinedFormData.set(name, data.getAll(name).join(", "));
    }
  }

  const steps = widget.get("steps") as Widget[];
  const widgets = steps.flatMap(s => s.widgets());
  const includeEmptyAnswers = widget.get("includeEmptyAnswers");
  const inputs: InputValidationElement[] = Array.from(
    form.querySelectorAll("input, select, textarea")
  );
  const inputNames = uniq(map(inputs, i => i.name));
  const reviewData: ReviewContent = [];

  for (const key of inputNames) {
    const answer = joinedFormData.get(key) as string;
    if (isEmpty(answer) && !includeEmptyAnswers) {
      // do not show empty answers
      continue;
    }
    // check if is hidden
    const input = inputs.find(i => i.name == key);
    if (
      key == "fax" ||
      (input &&
        input.type == "hidden" &&
        !input.classList.contains("show-in-review"))
    ) {
      // do not show hidden inputs
      continue;
    }
    // get title && step number
    const widget = widgets.find(w => getFieldName(w) == key);
    if (widget) {
      const step = steps.find(s =>
        s.widgets().find(w => getFieldName(w) == key)
      );
      const stepNumber = step?.get("stepNumber") as number;
      const title = (widget.get("label") ||
        widget.get("title") ||
        "") as string;

      if (!reviewData[stepNumber]) {
        reviewData[stepNumber] = [];
      }
      reviewData[stepNumber].push({
        title: title,
        value: getAnswerValue(answer, widget)
      });
    }
  }
  return reviewData;
}

function getAnswerValue(answer: string, widget: Widget): string {
  const emptyValue = "-";
  if (!answer || answer.length <= 0) {
    return emptyValue;
  }
  const className = widget.objClass();
  if (className != "FormDateWidget") {
    return answer;
  }
  const type = widget.get("dateType");
  if (type == "date") {
    return new Date(answer).toLocaleDateString();
  } else if (type == "datetime-local") {
    return new Date(answer).toLocaleString();
  }
  return answer;
}
