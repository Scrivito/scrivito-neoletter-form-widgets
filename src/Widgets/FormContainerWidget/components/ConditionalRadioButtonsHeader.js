import * as React from "react";
import * as Scrivito from "scrivito";
import { getFieldName } from "../utils/getFieldName";
import { SelectItem } from "./SelectComponent";
import { Mandatory } from "./MandatoryComponent";
import { HelpText } from "./HelpTextComponent";

export const RadioButtonsHeader = Scrivito.connect(
  ({ widget, onChangeSelected }) => {
    return (
      <div
        className={`mb-3 select-container ${
          Scrivito.isInPlaceEditingActive() ? "condition-wrapper" : ""
        }`}
      >
        <div className="select-title">
          <span className="text-super">{widget.get("title")}</span>
          {widget.get("required") && <Mandatory />}
          {widget.get("helpText") && <HelpText widget={widget} />}
        </div>
        <div className="row ">
          {widget.get("conditions").map((condition, index) => (
            <SelectItem
              selectionType={"radio"}
              name={getFieldName(widget)}
              onChange={onChangeSelected}
              value={condition.get("title")}
              id={condition.id()}
              required={widget.get("required")}
              key={index}
            />
          ))}
        </div>
      </div>
    );
  }
);
