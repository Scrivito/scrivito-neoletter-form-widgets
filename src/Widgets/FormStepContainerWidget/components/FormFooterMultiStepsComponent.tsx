import * as React from "react";
import * as Scrivito from "scrivito";
import { Review } from "./ReviewComponent";
import { prepareReviewContent } from "../utils/prepareReviewContent";
import { ReviewContent } from "../../../../types/types";
interface FormFooterMultiStepsProps {
  widget: Scrivito.Widget;
  onPageChange: (forward: boolean) => void;
  onSubmit: React.MouseEventHandler;
  currentStep: number;
  isLastPage: boolean;
  stepsLength: number;
  showReview: boolean;
}

export const FormFooterMultiSteps: React.FC<FormFooterMultiStepsProps> =
  Scrivito.connect(
    ({
      widget,
      onPageChange,
      onSubmit,
      currentStep,
      isLastPage,
      stepsLength,
      showReview
    }) => {
      const [show, setShow] = React.useState(false);
      const [reviewContent, setReviewContent] = React.useState<ReviewContent>(
        []
      );
      const doShowReview = isLastPage && showReview;
      return (
        <>
          <div className="form-buttons">
            <button
              className="btn btn-primary backward-button"
              onClick={() => onPageChange(false)}
              hidden={currentStep == 1 && !Scrivito.isInPlaceEditingActive()}>
              {widget.get("backwardButtonText") as string}
            </button>
            <div className="step-counter">
              {currentStep + " / " + stepsLength}
            </div>
            {doShowReview && (
              <button
                className="btn btn-primary review-button"
                onClick={() => onShowReview(widget, setReviewContent, setShow)}>
                {widget.get("reviewButtonText") as string}
              </button>
            )}
            <button
              className="btn btn-primary forward-button"
              onClick={isLastPage ? onSubmit : () => onPageChange(true)}>
              {isLastPage
                ? (widget.get("submitButtonText") as string)
                : (widget.get("forwardButtonText") as string)}
            </button>
          </div>
          {doShowReview && show && (
            <Review
              widget={widget}
              reviewContent={reviewContent}
              onHide={() => setShow(false)}
            />
          )}
        </>
      );
    }
  );

function onShowReview(
  widget: Scrivito.Widget,
  setReviewContent: (content: ReviewContent) => void,
  setShow: (show: boolean) => void
) {
  const reviewContent = prepareReviewContent(widget);

  setReviewContent(reviewContent);
  setShow(true);
}
