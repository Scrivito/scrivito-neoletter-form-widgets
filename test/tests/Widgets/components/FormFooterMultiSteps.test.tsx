import * as React from "react";
import * as Scrivito from "scrivito";
import { render, fireEvent } from "@testing-library/react";
import { FormFooterMultiSteps } from "../../../../src/Widgets/FormStepContainerWidget/components/FormFooterMultiStepsComponent";
import { DummyWidget } from "../../../helpers/dummyWidget";
import { prepareReviewContent } from "../../../../src/Widgets/FormStepContainerWidget/utils/prepareReviewContent";
import renderer from "react-test-renderer";

jest.mock(
  "../../../../src/Widgets/FormStepContainerWidget/utils/prepareReviewContent"
);

describe("FormFooterMultiSteps", () => {
  const mockPrepareReviewContent = jest.fn(() => [
    [{ title: "Field 1", value: "Value 1" }],
    [{ title: "Field 2", value: "Value 2" }]
  ]);

  const widget = new DummyWidget({
    backwardButtonText: "Back",
    reviewButtonText: "Review",
    submitButtonText: "Submit",
    forwardButtonText: "Forward"
  }) as unknown as Scrivito.Widget;

  afterEach(() => {
    (prepareReviewContent as jest.Mock).mockImplementation(
      mockPrepareReviewContent
    );
  });

  it("renders correctly with required buttons", () => {
    const { getByText } = render(
      <FormFooterMultiSteps
        widget={widget}
        onPageChange={() => {}}
        onSubmit={() => {}}
        currentStep={1}
        isLastPage={false}
        stepsLength={3}
        showReview={true}
      />
    );

    expect(getByText("Back")).toBeInTheDocument();
    expect(getByText("1 / 3")).toBeInTheDocument();
    expect(getByText("Forward")).toBeInTheDocument();
  });

  it("triggers review display correctly when Review button is clicked", () => {
    const { getByText } = render(
      <FormFooterMultiSteps
        widget={widget}
        onPageChange={() => {}}
        onSubmit={() => {}}
        currentStep={1}
        isLastPage={true}
        stepsLength={3}
        showReview={true}
      />
    );

    const reviewButton = getByText("Review");
    fireEvent.click(reviewButton);

    expect(mockPrepareReviewContent).toHaveBeenCalledWith(widget);
  });

  it("displays Submit button on last page", () => {
    const { getByText } = render(
      <FormFooterMultiSteps
        widget={widget}
        onPageChange={() => {}}
        onSubmit={() => {}}
        currentStep={3}
        isLastPage={true}
        stepsLength={3}
        showReview={true}
      />
    );

    expect(getByText("Submit")).toBeInTheDocument();
  });

  it("hides backward-button on the first page", () => {
    const { getByText } = render(
      <FormFooterMultiSteps
        widget={widget}
        onPageChange={() => {}}
        onSubmit={() => {}}
        currentStep={1}
        isLastPage={false}
        stepsLength={3}
        showReview={true}
      />
    );

    const backButton = getByText("Back");
    expect(backButton).toHaveAttribute("hidden");
  });

  it("shows backward-button on page 2", () => {
    const { getByText } = render(
      <FormFooterMultiSteps
        widget={widget}
        onPageChange={() => {}}
        onSubmit={() => {}}
        currentStep={2}
        isLastPage={false}
        stepsLength={3}
        showReview={true}
      />
    );

    const backButton = getByText("Back");
    expect(backButton).not.toHaveAttribute("hidden");
  });

  it("does not show Review button on last page if showReview is false", () => {
    const { queryByText } = render(
      <FormFooterMultiSteps
        widget={widget}
        onPageChange={() => {}}
        onSubmit={() => {}}
        currentStep={3}
        isLastPage={true}
        stepsLength={3}
        showReview={false}
      />
    );

    expect(queryByText("Review")).not.toBeInTheDocument();
  });

  it("renders correctly", () => {
    const tree = renderer
      .create(
        <FormFooterMultiSteps
          widget={widget}
          onPageChange={() => {}}
          onSubmit={() => {}}
          currentStep={1}
          isLastPage={true}
          stepsLength={3}
          showReview={true}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
