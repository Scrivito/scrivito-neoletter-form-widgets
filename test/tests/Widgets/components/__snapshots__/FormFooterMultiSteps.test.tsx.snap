// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFooterMultiSteps renders correctly 1`] = `
<div
  className="form-buttons"
>
  <button
    className="btn btn-primary backward-button"
    hidden={true}
    onClick={[Function]}
  >
    Back
  </button>
  <div
    className="step-counter"
  >
    1 / 3
  </div>
  <button
    className="btn btn-primary review-button"
    onClick={[Function]}
  >
    Review
  </button>
  <button
    className="btn btn-primary forward-button"
    disabled={false}
    onClick={[Function]}
  >
    Submit
  </button>
</div>
`;
